sympy/solvers/solveset.py

```diff
@@ -714,120 +514,96 @@ def _solve_abs(f, symbol):
+        expr_set_pairs = f.as_expr_set_pairs()
+        for (expr, in_set) in expr_set_pairs:
+            if in_set.is_Relational:
+                in_set = in_set.as_set()
+            if in_set.is_Interval:
+                dom -= in_set
+            solns = solver(expr, symbol, in_set)
+            result += solns
     else:
-        lhs, rhs_s = invert_complex(f, 0, symbol)
+        lhs, rhs_s = inverter(f, 0, symbol)
         if lhs == symbol:
+            # do some very minimal simplification since
+            # repeated inversion may have left the result
+            # in a state that other solvers (e.g. poly)
+            # would have simplified; this is done here
+            # rather than in the inverter since here it
+            # is only done once whereas there it would
+            # be repeated for each step of the inversion
+            if isinstance(rhs_s, FiniteSet):
+                rhs_s = FiniteSet(*[Mul(*
+                    signsimp(i).as_content_primitive())
+                    for i in rhs_s])
             result = rhs_s
         elif isinstance(rhs_s, FiniteSet):
-            equations = [lhs - rhs for rhs in rhs_s]
-            result = EmptySet()
-            for equation in equations:
+            for equation in [lhs - rhs for rhs in rhs_s]:
                 if equation == f:
                     if any(_has_rational_power(g, symbol)[0]
                            for g in equation.args) or _has_rational_power(
                            equation, symbol)[0]:
                         result += _solve_radical(equation,
                                                  symbol,
-                                                 solveset_complex)
+                                                 solver)
+                    elif equation.has(Abs):
+                        result += _solve_abs(f, symbol, domain)
                     else:
                         result += _solve_as_rational(equation, symbol,
-                                                 solveset_solver=solveset_complex,
-                                                 as_poly_solver=_solve_as_poly_complex)
+                                                     solveset_solver=solver,
+                                                     as_poly_solver=poly_solver)
                 else:
-                    result += solveset_complex(equation, symbol)
+                    result += solver(equation, symbol)
         else:
-            result = ConditionSet(symbol, Eq(f, 0), S.Complexes)
+            result = ConditionSet(symbol, Eq(f, 0), domain)
 
-    if isinstance(result, FiniteSet):
-        result = [s for s in result
-                  if isinstance(s, RootOf)
-                  or domain_check(original_eq, symbol, s)]
-        return FiniteSet(*result)
-    else:
-        return result
+    return result
 
 
-def solveset(f, symbol=None, domain=S.Complexes):
+def solveset(f, symbol=None, domain=S.Complexes, check=True, _first=True, ):
```

Comment:
These are private flags; they could be hidden in `kwargs` but there is no need. They aren't discussed as a use option and (now) the `check` flag reads `_check`. solveset is the front door to all the routines. It's just my preference to handle the helper (for check and first) within the function rather than in "yet another helper function". If you or anyone else feels strongly about not doing so, I can refactor.